mypy/checker.py

```diff
@@ -186,6 +187,13 @@ def visit_file(self, file_node: MypyFile, path: str) -> None:
                 self.fail(messages.ALL_MUST_BE_SEQ_STR.format(str_seq_s, all_s),
                           all_.node)
 
+    @contextmanager
+    def new_scope(self) -> Iterator[ConditionalTypeBinder]:
+        old_binder = self.binder
+        self.binder = ConditionalTypeBinder()
```

Comment:
This is tricky.  With nested functions / generator comprehensions, you need to lose the conditional information; for example,
```
x = 1 # type: Union[int, str]
x = 1
def foo() -> None:
    x + 1
x = 'a'
foo() # Exception!
```
But yeah, this doesn't work great for list/dict/set comprehensions.  Doing that right is a bit annoying, since the binder keeps track of literal expressions like (Var x), and [x for x in y] creates a new Var object (Var x) separate from the one in the existing scope.  Maybe I should explicitly look at the loop indices and just clear those?